### PR TITLE
Optimize index creation time

### PR DESCRIPTION
**Pull request by Ma-cat**
### What problem does this PR solve?
* Use loser tree to replace normal heap
* Optimize io and use mmap to replace fread
* Reconstruct merge and predic threads using ring queues
* Reconstruct merge and output threads using output queues
* Adjust the intermediate file output format and reduce the size of the intermediate file
* Reconstruct the locks between merge, predic, and output threads to reduce lock granularity

### Type of change
- [x] Refactoring
- [x] Performance Improvement
- [x] Test cases

### CODE REVIEW

File: src/storage/invertedindex/common/external_sort_merger.cpp [hunk 1]
```diff
@@ -208,18 +221,13 @@ void SortMerger<KeyType, LenType>::Init(DirectIO &io_stream) {
 
 template <typename KeyType, typename LenType>
 void SortMerger<KeyType, LenType>::Predict(DirectIO &io_stream) {
+    UniquePtr<char[]> data_buf = MakeUnique<char[]>(PRE_BUF_SIZE_);
```
Comment (yuzhichang) [inline src/storage/invertedindex/common/external_sort_merger.cpp:224]: Change `pre_heap_` something like:
    std::priority_queue<KeyAddr, KeyAddr::>  > pre_heap_;   //!< predict heap

File: src/storage/invertedindex/column_inverter.cpp [hunk 2]
```diff
@@ -299,4 +302,133 @@ void ColumnInverter::SpillSortResults(FILE *spill_file, u64 &tuple_count) {
...
+    next_start_offset = ftell(spill_file);
+    data_size = next_start_offset - data_start_offset;
+    fseek(spill_file, data_size_pos, SEEK_SET);
+    fwrite(&data_size, sizeof(u32), 1, spill_file); // update offset for next spill
+    fseek(spill_file, next_start_offset_pos, SEEK_SET);
+    fwrite(&next_start_offset, sizeof(u64), 1, spill_file);
+    fseek(spill_file, next_start_offset, SEEK_SET);
+}
+
+void ColumnInverter::SpillSortResults(FILE *spill_file, u64 &tuple_count, UniquePtr<BufWriter>& buf_writer) {
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/column_inverter.cpp:260]: now olny keep `void ColumnInverter::SpillSortResults(FILE *spill_file, u64 &tuple_count, UniquePtr<BufWriter>& buf_writer)`

File: src/storage/invertedindex/common/external_sort_merger.cpp [hunk 3]
```diff
@@ -394,31 +451,193 @@ void SortMerger<KeyType, LenType>::Run() {
...
+                    std::unique_lock lock(out_queue_mtx_);
+                    term_tuple_list_queue_.push(std::move(tuple_list));
+                    out_queue_con_.notify_one();
+                }
+                tuple_list = MakeUnique<TermTupleList>(out_key.term_);
+                tuple_list->Add(out_key.doc_id_, out_key.term_pos_);
+            } else {
+                tuple_list->Add(out_key.doc_id_, out_key.term_pos_);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cpp:492]: now just use a `tuple_list->Add(out_key.doc_id_, out_key.term_pos_);` outside `if-else`

File: src/storage/invertedindex/common/buf_writer.cppm [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace infinity {
+
+export struct BufWriter {
+    BufWriter(FILE *spill_file, SizeT spill_buf_size) : spill_file_(spill_file), spill_buf_size_(spill_buf_size) {
+        spill_buffer_ = MakeUnique<char_t[]>(spill_buf_size_);
+    }
+
+    void Write(const char* data, SizeT data_size) {
+        if (spill_buf_idx_ + data_size >= spill_buf_size_) {
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/buf_writer.cppm]: fixed

File: src/storage/invertedindex/common/buf_writer.cppm [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <cstring>
+#include <cstdio>
+
+export module buf_writer;
+import stl;
+
+namespace infinity {
+
+export struct BufWriter {
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/buf_writer.cppm:13]: added

File: src/storage/invertedindex/common/external_sort_merger.cpp [hunk 6]
```diff
@@ -17,11 +17,21 @@ module;
...
+#include <cstring>
+#include <sys/mman.h>
+#include <cstdio>
+#include <unistd.h>
+#include <fcntl.h>
 
 module external_sort_merger;
 
 import stl;
+import mmap;
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cpp]: removed

File: src/storage/invertedindex/common/external_sort_merger.cpp [hunk 7]
```diff
@@ -342,6 +361,44 @@ void SortMerger<KeyType, LenType>::Merge() {
     }
 }
 
+template <typename KeyType, typename LenType>
+void SortMerger<KeyType, LenType>::OutputByQueue(FILE *f) {
+    DirectIO io_stream(f, "w");
+    while (count_ > 0) {
+        // wait its turn to output
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cpp]: This function has not been used and has been deleted

File: src/storage/invertedindex/common/external_sort_merger.cppm [hunk 8]
```diff
@@ -206,11 +239,143 @@ struct KeyAddress<TermTuple, LenType> {
...
+        : total_buffers_(total_buffers), buffer_size_(buffer_size), head_(0), tail_(0), full_(false) {
+        buffer_array_.resize(total_buffers);
+        buffer_real_size_.resize(total_buffers);
+        buffer_real_num_.resize(total_buffers);
+        for (auto& buf : buffer_array_) {
+            buf = MakeUnique<char[]>(buffer_size);
+        }
+    }
+
+    void Put(const TermTupleList& tuple_list) {
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cppm]: This code has not been used and has been deleted

File: src/storage/invertedindex/common/external_sort_merger.cpp (hunk 1 above)
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cpp:224]: resolved

File: src/storage/invertedindex/common/loser_tree.cppm [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+#include <cassert>
+#include <functional>
+
+export module loser_tree;
+
+import stl;
+
+namespace infinity {
+
+export template <typename ValueType, typename Comparator = std::greater<ValueType>>
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/loser_tree.cppm:14]: resolved

File: src/storage/invertedindex/common/external_sort_merger.cppm [hunk 10]
```diff
@@ -271,31 +443,86 @@ class SortMerger {
...
-            RUN_BUF_SIZE_ = PRE_BUF_SIZE_ * MAX_GROUP_SIZE_;
-        if (OUT_BUF_SIZE_ < min_buff_size_required)
-            OUT_BUF_SIZE_ = min_buff_size_required;
-    }
+export template <typename KeyType, typename LenType>
+requires std::same_as<KeyType, TermTuple>
+class SortMergerTermTuple : public SortMerger<KeyType, LenType> {
+protected:
+    typedef SortMergerTermTuple<KeyType, LenType> self_t;
+    using Super = SortMerger<KeyType, LenType>;
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cppm:431]: remove some unused Super::members

File: src/storage/invertedindex/common/external_sort_merger.cpp [hunk 11]
```diff
@@ -312,23 +319,35 @@ void SortMerger<KeyType, LenType>::Merge() {
         // reach the end of a microrun
         if (micro_run_idx_[idx] == num_micro_run_[idx]) {
             IASSERT(micro_run_pos_[idx] <= size_micro_run_[idx]);
-            std::unique_lock lock(pre_buf_mtx_);
-            while (pre_buf_size_ == 0)
-                pre_buf_con_.wait(lock);
+            std::unique_lock lock(cycle_buf_mtx_);
+
+            cycle_buf_con_.wait(lock, [this]() {
+                return !this->cycle_buffer_->IsEmpty() || (this->read_finish_ && this->cycle_buffer_->IsEmpty());
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cpp]: resolved

File: src/storage/invertedindex/common/external_sort_merger.cpp [hunk 12]
```diff
@@ -232,58 +240,57 @@ void SortMerger<KeyType, LenType>::Predict(DirectIO &io_stream) {
...
-            LenType len = *(LenType *)(pre_buf_ + pos);
+            LenType len = *(LenType *)(data_ptr + pos);
             if (pos + sizeof(LenType) + len > s) {
-                // the last record of this microrun
                 IASSERT(last_pos != (u32)-1); // buffer too small that can't hold one record
-                len = *(LenType *)(pre_buf_ + last_pos) + sizeof(LenType);
+                len = *(LenType *)(data_ptr + last_pos) + sizeof(LenType);
                 char *tmp = (char *)malloc(len);
-                memcpy(tmp, pre_buf_ + last_pos, len);
+                memcpy(tmp, data_ptr + last_pos, len);
```
[the remark this answers is not in the document]
Comment (Ma-cat) [reply src/storage/invertedindex/common/external_sort_merger.cpp]: resolved

**Pull request merged**